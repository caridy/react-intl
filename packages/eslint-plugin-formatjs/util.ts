import {TSESTree} from '@typescript-eslint/typescript-estree';
import {Scope} from 'eslint';

export interface MessageDescriptor {
  id?: string;
  defaultMessage?: string;
  description?: string;
}

export interface MessageDescriptorNodeInfo {
  message: MessageDescriptor;
  messageNode?: TSESTree.Property['value'] | TSESTree.JSXAttribute['value'];
  messagePropNode?: TSESTree.Property | TSESTree.JSXAttribute;
  descriptionNode?: TSESTree.Property['value'] | TSESTree.JSXAttribute['value'];
  idValueNode?: TSESTree.Property['value'] | TSESTree.JSXAttribute['value'];
  idPropNode?: TSESTree.Property | TSESTree.JSXAttribute;
}

function isStringLiteral(node: TSESTree.Node): node is TSESTree.StringLiteral {
  return node.type === 'Literal' && typeof node.value === 'string';
}

function findReferenceImport(
  id: TSESTree.Identifier,
  importedVars: Scope.Variable[]
) {
  return importedVars.find(
    v => !!v.references.find(ref => ref.identifier === id)
  );
}

function staticallyEvaluateStringConcat(
  node: TSESTree.BinaryExpression
): [result: string, isStaticallyEvaluatable: boolean] {
  if (!isStringLiteral(node.right)) {
    return ['', false];
  }
  if (isStringLiteral(node.left)) {
    return [String(node.left.value) + node.right.value, true];
  }
  if (node.left.type === 'BinaryExpression') {
    const [result, isStaticallyEvaluatable] = staticallyEvaluateStringConcat(
      node.left
    );
    return [result + node.right.value, isStaticallyEvaluatable];
  }
  return ['', false];
}

function isIntlFormatMessageCall(node: TSESTree.Node) {
  return (
    node.type === 'CallExpression' &&
    node.callee.type === 'MemberExpression' &&
    node.callee.object.type === 'Identifier' &&
    node.callee.object.name === 'intl' &&
    node.callee.property.type === 'Identifier' &&
    node.callee.property.name === 'formatMessage' &&
    node.arguments.length >= 1 &&
    node.arguments[0].type === 'ObjectExpression'
  );
}

function isSingleMessageDescriptorDeclaration(
  id: TSESTree.LeftHandSideExpression,
  importedVars: Scope.Variable[]
) {
  if (id.type !== 'Identifier') {
    return false;
  }
  const importedVar = findReferenceImport(id, importedVars);
  if (!importedVar) {
    return false;
  }
  return importedVar.name === '_' || importedVar.name === 'defineMessage';
}
function isMultipleMessageDescriptorDeclaration(
  id: TSESTree.LeftHandSideExpression,
  importedVars: Scope.Variable[]
) {
  if (id.type !== 'Identifier') {
    return false;
  }
  const importedVar = findReferenceImport(id, importedVars);
  if (!importedVar) {
    return false;
  }
  return importedVar.name === 'defineMessages';
}

function extractMessageDescriptor(
  node?: TSESTree.Expression
): MessageDescriptorNodeInfo | undefined {
  if (!node || node.type !== 'ObjectExpression') {
    return;
  }
  const result: MessageDescriptorNodeInfo = {
    message: {},
    messageNode: undefined,
    messagePropNode: undefined,
    descriptionNode: undefined,
    idValueNode: undefined,
  };
  for (const prop of node.properties) {
    if (prop.type !== 'Property' || prop.key.type !== 'Identifier') {
      continue;
    }
    const valueNode = prop.value;
    let value: string | undefined = undefined;
    if (isStringLiteral(valueNode)) {
      value = valueNode.value;
    } else if (valueNode.type === 'BinaryExpression') {
      const [result, isStatic] = staticallyEvaluateStringConcat(valueNode);
      if (isStatic) {
        value = result;
      }
    }

    switch (prop.key.name) {
      case 'defaultMessage':
        result.messagePropNode = prop;
        result.messageNode = valueNode;
        result.message.defaultMessage = value;
        break;
      case 'description':
        result.descriptionNode = valueNode;
        result.message.description = value;
        break;
      case 'id':
        result.message.id = value;
        result.idValueNode = valueNode;
        result.idPropNode = prop;
        break;
    }
  }
  return result;
}

function extractMessageDescriptorFromJSXElement(
  node?: TSESTree.JSXOpeningElement
):
  | [MessageDescriptorNodeInfo, TSESTree.ObjectExpression | undefined]
  | undefined {
  if (!node || !node.attributes) {
    return;
  }
  let values: TSESTree.ObjectExpression | undefined;
  const result: MessageDescriptorNodeInfo = {
    message: {},
    messageNode: undefined,
    messagePropNode: undefined,
    descriptionNode: undefined,
    idValueNode: undefined,
    idPropNode: undefined,
  };
  for (const prop of node.attributes) {
    if (prop.type !== 'JSXAttribute' || prop.name.type !== 'JSXIdentifier') {
      continue;
    }
    const key = prop.name;
    let valueNode = prop.value;
    let value: string | undefined = undefined;
    if (valueNode) {
      if (isStringLiteral(valueNode)) {
        value = valueNode.value;
      } else if (
        valueNode?.type === 'JSXExpressionContainer' &&
        valueNode.expression.type === 'BinaryExpression'
      ) {
        const [result, isStatic] = staticallyEvaluateStringConcat(
          valueNode.expression
        );
        if (isStatic) {
          value = result;
        }
      }
    }

    switch (key.name) {
      case 'defaultMessage':
        result.messagePropNode = prop;
        result.messageNode = valueNode;
        if (value) {
          result.message.defaultMessage = value;
        }
        break;
      case 'description':
        result.descriptionNode = valueNode;
        if (value) {
          result.message.description = value;
        }
        break;
      case 'id':
        result.idValueNode = valueNode;
        result.idPropNode = prop;
        if (value) {
          result.message.id = value;
        }
        break;
      case 'values':
        if (
          valueNode?.type === 'JSXExpressionContainer' &&
          valueNode.expression.type === 'ObjectExpression'
        ) {
          values = valueNode.expression;
        }
        break;
    }
  }
  if (
    !result.messagePropNode &&
    !result.descriptionNode &&
    !result.idPropNode
  ) {
    return;
  }
  return [result, values];
}

function extractMessageDescriptors(node?: TSESTree.Expression) {
  if (!node || node.type !== 'ObjectExpression' || !node.properties.length) {
    return [];
  }
  const msgs = [];
  for (const prop of node.properties) {
    if (prop.type !== 'Property') {
      continue;
    }
    const msg = prop.value;
    if (msg.type !== 'ObjectExpression') {
      continue;
    }
    const nodeInfo = extractMessageDescriptor(msg);
    if (nodeInfo) {
      msgs.push(nodeInfo);
    }
  }
  return msgs;
}

export function extractMessages(
  node: TSESTree.Node,
  importedMacroVars: Scope.Variable[],
  excludeMessageDeclCalls = false
): Array<[MessageDescriptorNodeInfo, TSESTree.Expression | undefined]> {
  if (node.type === 'CallExpression') {
    const expr = node;
    const fnId = expr.callee;
    if (
      (!excludeMessageDeclCalls &&
        isSingleMessageDescriptorDeclaration(fnId, importedMacroVars)) ||
      isIntlFormatMessageCall(node)
    ) {
      const msgDescriptorNodeInfo = extractMessageDescriptor(expr.arguments[0]);
      if (msgDescriptorNodeInfo) {
        return [[msgDescriptorNodeInfo, expr.arguments[1]]];
      }
    } else if (
      !excludeMessageDeclCalls &&
      isMultipleMessageDescriptorDeclaration(fnId, importedMacroVars)
    ) {
      return extractMessageDescriptors(expr.arguments[0]).map(msg => [
        msg,
        undefined,
      ]);
    }
  } else if (
    node.type === 'JSXOpeningElement' &&
    node.name &&
    node.name.type === 'JSXIdentifier' &&
    node.name.name === 'FormattedMessage'
  ) {
    const msgDescriptorNodeInfo = extractMessageDescriptorFromJSXElement(node);
    if (msgDescriptorNodeInfo) {
      return [msgDescriptorNodeInfo];
    }
  }
  return [];
}
