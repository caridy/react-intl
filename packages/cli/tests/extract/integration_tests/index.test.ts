import {resolve, join} from 'path';
import {promisify} from 'util';
import {exec as nodeExec} from 'child_process';
import {readJSON, mkdirp, readdir} from 'fs-extra';
import * as _rimraf from 'rimraf';
const exec = promisify(nodeExec);
const rimraf = promisify(_rimraf);

const BIN_PATH = resolve(__dirname, '../../../bin/formatjs');
const ARTIFACT_PATH = resolve(__dirname, 'test_artifacts');

beforeEach(async () => {
  await mkdirp(join(__dirname, 'test_artifacts'));
  await rimraf(ARTIFACT_PATH);
});

test('basic case: help', async () => {
  const {stdout, stderr} = await exec(`${BIN_PATH} extract --help`);
  expect(stdout).toMatchSnapshot();
  expect(stderr).toBe('');
}, 20000);

test('basic case: defineMessages -> stdout', async () => {
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract ${join(__dirname, 'defineMessages/actual.js')}`
  );
  expect(JSON.parse(stdout)).toMatchSnapshot();
  expect(stderr).toBe('');
}, 20000);

test('[glob] basic case: defineMessages -> stdout', async () => {
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract ${join(__dirname, 'defineMessages/*.js')}`
  );
  expect(JSON.parse(stdout)).toMatchSnapshot();
  expect(stderr).toBe('');
}, 20000);

test('basic case: defineMessages -> directory', async () => {
  process.chdir(__dirname);
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract defineMessages/actual.js --messages-dir ${ARTIFACT_PATH}`
  );
  expect(stdout).toBe('');
  expect(stderr).toBe('');
  // Write to test_artifacts/defineMessages/actual.json
  expect(
    await readdir(join(ARTIFACT_PATH, 'defineMessages'))
  ).toMatchSnapshot();
  expect(
    await readJSON(join(ARTIFACT_PATH, 'defineMessages/actual.json'))
  ).toMatchSnapshot();
}, 20000);

test('basic case: defineMessages -> directory with location', async () => {
  process.chdir(__dirname);
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract defineMessages/actual.js --extract-source-location --messages-dir ${ARTIFACT_PATH}`
  );
  expect(stdout).toBe('');
  expect(stderr).toBe('');
  // Write to test_artifacts/defineMessages/actual.json
  expect(
    await readdir(join(ARTIFACT_PATH, 'defineMessages'))
  ).toMatchSnapshot();
  expect(
    await readJSON(join(ARTIFACT_PATH, 'defineMessages/actual.json'))
  ).toMatchSnapshot();
}, 20000);

test('typescript -> stdout', async () => {
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract ${join(__dirname, 'typescript/actual.tsx')}`
  );
  expect(JSON.parse(stdout)).toMatchSnapshot();
  expect(stderr).toBe('');
}, 20000);

test('ignore -> stdout', async () => {
  const ignore = "--ignore '*.ignore.*'";

  const tsResult = await exec(
    `${BIN_PATH} extract ${join(__dirname, 'typescript/actual.tsx')} ${ignore}`
  );
  expect(JSON.parse(tsResult.stdout)).toMatchSnapshot();
  expect(tsResult.stderr).toBe('');

  const jsResult = await exec(
    `${BIN_PATH} extract ${join(__dirname, 'defineMessages/*.js')} ${ignore}`
  );
  expect(JSON.parse(jsResult.stdout)).toMatchSnapshot();
  expect(jsResult.stderr).toBe('');
}, 20000);

test('duplicated descriptor ids shows warning', async () => {
  const {stderr, stdout} = await exec(
    `${BIN_PATH} extract ${join(__dirname, 'duplicated/*.tsx')}`
  );
  expect(JSON.parse(stdout)).toMatchSnapshot();
  expect(stderr).toContain('Duplicate message id: "foo"');
}, 20000);

test('duplicated descriptor ids throws', async () => {
  expect(async () => {
    await exec(
      `${BIN_PATH} extract --throws ${join(__dirname, 'duplicated/*.tsx')}`
    );
  }).rejects.toThrowError('Duplicate message id: "foo"');
}, 20000);
