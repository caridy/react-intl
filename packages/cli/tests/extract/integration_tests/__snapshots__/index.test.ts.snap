// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[glob] basic case: defineMessages -> stdout 1`] = `
Object {
  "ae494": Object {
    "defaultMessage": "No ID",
    "description": "no ID",
  },
  "app.home.kittens": Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
  },
  "escaped.apostrophe": Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
  },
  "foo.bar.baz": Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
  },
  "foo.bar.biff": Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
  },
  "ignore": Object {
    "defaultMessage": "ignore",
  },
  "inline-msg": Object {
    "defaultMessage": "inline message",
    "description": "inline description",
  },
  "trailing.ws": Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
  },
}
`;

exports[`basic case: defineMessages -> out-file 1`] = `
Object {
  "ae494": Object {
    "defaultMessage": "No ID",
    "description": "no ID",
  },
  "app.home.kittens": Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
  },
  "escaped.apostrophe": Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
  },
  "foo.bar.baz": Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
  },
  "foo.bar.biff": Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
  },
  "inline-msg": Object {
    "defaultMessage": "inline message",
    "description": "inline description",
  },
  "trailing.ws": Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
  },
}
`;

exports[`basic case: defineMessages -> out-file with location 1`] = `
Object {
  "ae494": Object {
    "col": 18,
    "defaultMessage": "No ID",
    "description": "no ID",
    "end": 937,
    "file": "defineMessages/actual.js",
    "line": 34,
    "start": 876,
  },
  "app.home.kittens": Object {
    "col": 10,
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "end": 522,
    "file": "defineMessages/actual.js",
    "line": 15,
    "start": 372,
  },
  "escaped.apostrophe": Object {
    "col": 10,
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "end": 786,
    "file": "defineMessages/actual.js",
    "line": 25,
    "start": 662,
  },
  "foo.bar.baz": Object {
    "col": 9,
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "end": 246,
    "file": "defineMessages/actual.js",
    "line": 5,
    "start": 141,
  },
  "foo.bar.biff": Object {
    "col": 10,
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "end": 360,
    "file": "defineMessages/actual.js",
    "line": 10,
    "start": 258,
  },
  "inline-msg": Object {
    "col": 10,
    "defaultMessage": "inline message",
    "description": "inline description",
    "end": 1416,
    "file": "defineMessages/actual.js",
    "line": 53,
    "start": 1269,
  },
  "trailing.ws": Object {
    "col": 21,
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "end": 650,
    "file": "defineMessages/actual.js",
    "line": 20,
    "start": 545,
  },
}
`;

exports[`basic case: defineMessages -> stdout 1`] = `
Object {
  "ae494": Object {
    "defaultMessage": "No ID",
    "description": "no ID",
  },
  "app.home.kittens": Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
  },
  "escaped.apostrophe": Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
  },
  "foo.bar.baz": Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
  },
  "foo.bar.biff": Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
  },
  "inline-msg": Object {
    "defaultMessage": "inline message",
    "description": "inline description",
  },
  "trailing.ws": Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
  },
}
`;

exports[`basic case: help 1`] = `
"Usage: formatjs extract [options] [files...]

Extract string messages from React components that use react-intl.
The input language is expected to be TypeScript or ES2017 with JSX.

Options:
  --out-file <path>                                     The target file path where the plugin will output an aggregated \`.json\` file of allthe translations from the \`files\` supplied.
  This flag will ignore --messages-dir
  --id-interpolation-pattern <pattern>                  If certain message descriptors don't have id, this \`pattern\` will be used to automaticallygenerate IDs for them. Default to \`[contenthash:5]\`.
  See https://github.com/webpack/loader-utils#interpolatename for sample patterns (default: \\"[contenthash:5]\\")
  --extract-source-location                             Whether the metadata about the location of the message in the source file should be extracted. If \`true\`, then \`file\`, \`start\`, and \`end\` fields will exist for each extracted message descriptors. (default: false)
  --remove-default-message                              Remove \`defaultMessage\` field in generated js after extraction (default: false)
  --additional-component-names <comma-separated-names>  Additional component names to extract messages from, e.g: \`['FormattedFooBarMessage']\`. **NOTE**: By default we check for the fact that \`FormattedMessage\` is imported from \`moduleSourceName\` to make sure variable alias works. This option does not do that so it's less safe.
  --extract-from-format-message-call                    Opt-in to extract from \`intl.formatMessage\` call with the same restrictions, e.g: has to be called with object literal such as \`intl.formatMessage({ id: 'foo', defaultMessage: 'bar', description: 'baz'})\` (default: true)
  --ignore <files>                                      List of glob paths to **not** extract translations from.
  --throws                                              Whether to throw an exception when we fail to process any file in the batch.
  --pragma <pragma>                                     parse specific additional custom pragma. This allows you to tag certain file with metadata such as \`project\`. For example with this file:
  
        \`\`\`
        // @intl-meta project:my-custom-project
        import {FormattedMessage} from 'react-intl';
  
        <FormattedMessage defaultMessage=\\"foo\\" id=\\"bar\\" />;
        \`\`\`
  
        and with option \`{pragma: \\"@intl-meta\\"}\`, we'll parse out \`// @intl-meta project:my-custom-project\` into \`{project: 'my-custom-project'}\` in the result file.
  -h, --help                                            display help for command
"
`;

exports[`duplicated descriptor ids shows warning 1`] = `
Object {
  "foo": Object {
    "defaultMessage": "Bar",
  },
}
`;

exports[`ignore -> stdout JS 1`] = `
Object {
  "ae494": Object {
    "defaultMessage": "No ID",
    "description": "no ID",
  },
  "app.home.kittens": Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
  },
  "escaped.apostrophe": Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
  },
  "foo.bar.baz": Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
  },
  "foo.bar.biff": Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
  },
  "ignore": Object {
    "defaultMessage": "ignore",
  },
  "inline-msg": Object {
    "defaultMessage": "inline message",
    "description": "inline description",
  },
  "trailing.ws": Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
  },
}
`;

exports[`ignore -> stdout TS 1`] = `
Object {
  "ae494": Object {
    "defaultMessage": "No ID",
    "description": "no ID",
  },
  "app.home.kittens": Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
  },
  "c63ed": Object {
    "defaultMessage": "No Desc",
  },
  "escaped.apostrophe": Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
  },
  "foo.bar.baz": Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
  },
  "foo.bar.biff": Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
  },
  "inline": Object {
    "defaultMessage": "formatted message",
    "description": "foo",
  },
  "trailing.ws": Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
  },
}
`;

exports[`typescript -> stdout 1`] = `
Object {
  "ae494": Object {
    "defaultMessage": "No ID",
    "description": "no ID",
  },
  "app.home.kittens": Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
  },
  "c63ed": Object {
    "defaultMessage": "No Desc",
  },
  "escaped.apostrophe": Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
  },
  "foo.bar.baz": Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
  },
  "foo.bar.biff": Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
  },
  "inline": Object {
    "defaultMessage": "formatted message",
    "description": "foo",
  },
  "trailing.ws": Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
  },
}
`;
